### PR TITLE
[FLINK-22639][runtime] ClassLoaderUtil cannot print classpath of Flin…

### PR DESCRIPTION
**Pull request by paul8263**
## What is the purpose of the change

Fixed missing classloader URL when using ClassLoaderUtil.getUserCodeClassLoaderInfo to retrieve the info of SafetyNetWrapperClassLoader.

## Brief change log

  - Added overridden getURLs method in SafetyNetWrapperClassLoader
  - Added corresponding unit tests
  - Refactored getUserCodeClassLoaderInfo in ClassLoaderUtil, by extracting the URL formatting logic to an independent method.

## Verifying this change

This change is already covered by existing tests, such as:

 - FlinkUserCodeClassLoadersTest.testGetClassLoaderInfo
 - FlinkUserCodeClassLoadersTest.testGetClassLoaderInfoWithClassLoaderClosed

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

**Comment by paul8263** [comment]
I noticed that there were several code format issues and those were fixed in 760d7a3.

File: flink-runtime/src/test/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoadersTest.java [hunk 1]
```diff
@@ -208,6 +212,64 @@ public void testRepeatedParentFirstPatternClass() throws Exception {
...
+                        NOOP_EXCEPTION_HANDLER,
+                        true);
+
+        String formattedURL = null;
+        try {
+            formattedURL = ClassLoaderUtil.formatURL(childCodePath);
+        } catch (IOException e) {
+            // Should not fail here
+            fail();
+        }
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoadersTest.java:238]:
```suggestion
        String formattedURL = ClassLoaderUtil.formatURL(childCodePath);
```
Let the exception bubble up

File: flink-runtime/src/test/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoadersTest.java [hunk 2]
```diff
@@ -208,6 +212,64 @@ public void testRepeatedParentFirstPatternClass() throws Exception {
         childClassLoader.close();
     }
 
+    @Test
+    public void testGetClassLoaderInfo() {
+        final String className = FlinkUserCodeClassLoadersTest.class.getName();
+        final String parentFirstPattern = className.substring(0, className.lastIndexOf('.'));
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoadersTest.java:218]: why is this necessary?

File: flink-runtime/src/test/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoadersTest.java [hunk 3]
```diff
@@ -208,6 +212,64 @@ public void testRepeatedParentFirstPatternClass() throws Exception {
...
+                        parentClassLoader,
+                        new String[] {parentFirstPattern},
+                        NOOP_EXCEPTION_HANDLER,
+                        true);
+
+        try {
+            childClassLoader.close();
+        } catch (IOException e) {
+            fail("Cannot close child classloader");
+        }
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoadersTest.java:266]: same as above

File: flink-runtime/src/test/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoadersTest.java [hunk 4]
```diff
@@ -208,6 +212,64 @@ public void testRepeatedParentFirstPatternClass() throws Exception {
...
+
+        assertTrue(
+                ClassLoaderUtil.getUserCodeClassLoaderInfo(childClassLoader)
+                        .startsWith("Cannot access classloader info due to an exception."));
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoadersTest.java:270]: use hamcrest matcher instead; assertThat(...).(StringStartsWith.startsWith(...); it provides much better error messages

**Review by zentol — CHANGES_REQUESTED** [review]

**Comment by paul8263** [comment]
Thank you @zentol . I reworked those codes in ce1bf53 .

File: flink-runtime/src/test/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoadersTest.java [hunk 5]
```diff
@@ -208,6 +212,50 @@ public void testRepeatedParentFirstPatternClass() throws Exception {
...
+                        NOOP_EXCEPTION_HANDLER,
+                        true);
+
+        String formattedURL = ClassLoaderUtil.formatURL(childCodePath);
+
+        assertEquals(
+                ClassLoaderUtil.getUserCodeClassLoaderInfo(childClassLoader),
+                "URL ClassLoader:" + formattedURL);
+
+        childClassLoader.close();
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoadersTest.java:235]: if you want to ensure that this CL is closed then you should use a finally block to ensure it is closed if the test fails

File: flink-runtime/src/test/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoadersTest.java [hunk 6]
```diff
@@ -208,6 +212,50 @@ public void testRepeatedParentFirstPatternClass() throws Exception {
...
+                        parentClassLoader,
+                        new String[] {},
+                        NOOP_EXCEPTION_HANDLER,
+                        true);
+
+        childClassLoader.close();
+
+        assertThat(
+                ClassLoaderUtil.getUserCodeClassLoaderInfo(childClassLoader),
+                is(startsWith("Cannot access classloader info due to an exception.")));
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoadersTest.java:256]:
```suggestion
                startsWith("Cannot access classloader info due to an exception."));
```
I would think that `is` isn't necessary.

**Review by zentol — CHANGES_REQUESTED** [review]
2 more smaller things

**Comment by paul8263** [comment]
@zentol I think there might not be necessary to ensure that the classloader is closed so I kept it the same as other test cases.
Also I noticed the method `createChildFirstClassLoader` already exists in `FlinkUserCodeClassLoadersTest`. I used it to keep the code clean. Is there anything else I can help?

File: flink-runtime/src/test/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoadersTest.java (hunk 1 above)
Comment (paul8263) [reply flink-runtime/src/test/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoadersTest.java]: It was fixed in  c927703

File: flink-runtime/src/test/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoadersTest.java (hunk 3 above)
Comment (paul8263) [reply flink-runtime/src/test/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoadersTest.java]: It was fixed in c927703
